### PR TITLE
New ztest API

### PR DESCRIPTION
**Pull request by oyvindronningstad**

### CODE REVIEW

**Review by Vge0rge — APPROVED** [review]

**Comment by gopiotr** [comment]
Please do not merge it yet. I'm in the middle of review - I'll try to finish it today

**Review by Vge0rge — APPROVED** [review]

**Comment by oyvindronningstad** [comment]
@gopiotr Thanks. Please remove the DNM when you're done.

File: tests/drivers/fprotect/positive/src/main.c [hunk 1]
```diff
@@ -23,7 +23,7 @@ static uint32_t control_read_addr = 0x6032;
 static uint8_t read_data_before[sizeof(write_data)];
 static uint8_t read_data_after[sizeof(write_data)];
 
-static void test_flash_write(void)
+ZTEST(test_fprotect_positive, test_flash_awrite)
```
Comment (gopiotr) [inline tests/drivers/fprotect/positive/src/main.c:26]: I'm not a big fan of this solution. It's quite tricky to name it as "awrite" only to enforce test order. If you really need to do this in this way, please name tests in more meaningful way, like `test_1_...`, `test_2_...`, etc.  In this case please move this `flash_write` to test suite setup, as for example:
```c
static void *setup(void)
{
	int retval = flash_read(flash_dev, control_read_addr, read_data_before,
				ARRAY_SIZE(read_data_before));

	zassert_true(retval == 0, "flash_read failed");
	retval = flash_write(flash_dev, valid_write_addr, write_data,
			     ARRAY_SIZE(write_data));
	zassert_true(retval == 0, "flash_write failed");

	return NULL;
}

ZTEST_SUITE(test_fprotect_positive, NULL, setup, NULL, NULL, NULL);
```

File: tests/drivers/fprotect/negative/src/main.c [hunk 2]
```diff
@@ -88,19 +88,19 @@ static void test_flash_write_protected_fails(void)
...
-static void test_flash_write_protected_unmodified(void)
+ZTEST(test_fprotect_negative, test_flash_write_protected_unmodified)
 {
 	flash_write_protected_unmodified(TEST_FPROTECT_WRITE_ADDR);
 }
 
-static void test_bootloader_protection(void)
+ZTEST(test_fprotect_negative, test_bootloader_protection)
 {
 #ifdef CONFIG_SECURE_BOOT
```
Comment (gopiotr) [inline tests/drivers/fprotect/negative/src/main.c:98]: You can consider using in this test `Z_TEST_SKIP_IFNDEF`:
https://github.com/zephyrproject-rtos/zephyr/blob/bb6fa5707f77d72307008405f19dd83247405392/subsys/testsuite/ztest/include/zephyr/ztest_test_new.h#L370-L379

Thanks to that, this testcase will be marked as SKIP when there is no `CONFIG_SECURE_BOOT` enabled. Without marking it explicate as skip, this test will be marked as PASS each time when there is no `SECURE_BOOT`, what could be misleading.

File: tests/drivers/fprotect/negative/src/main.c [hunk 3]
```diff
@@ -88,19 +88,19 @@ static void test_flash_write_protected_fails(void)
...
 	flash_write_protected_fails(TEST_FPROTECT_BOOTLOADER_PROTECTED, true);
 #endif
 }
 
-static void test_flash_read_protected_fails_r(void)
+ZTEST(test_fprotect_negative, test_flash_read_protected_fails_r)
 {
 #ifdef CONFIG_HAS_HW_NRF_ACL
```
Comment (gopiotr) [inline tests/drivers/fprotect/negative/src/main.c:105]: The same as in previous test - please try to use `Z_TEST_SKIP_IFNDEF`

**Review by gopiotr — CHANGES_REQUESTED** [review]

File: tests/drivers/fprotect/positive/src/main.c (hunk 1 above)
Comment (oyvindronningstad) [reply tests/drivers/fprotect/positive/src/main.c:26]: I think it should stay as a test. It is a test, it just has side effects that affect the next two tests. It's a test that deals with flash, after all.

**Comment by oyvindronningstad** [comment]
@gopiotr Thanks, I have addressed all your comments.

**Comment by gopiotr** [comment]
@oyvindronningstad thank you for your work.

I tried to run all those tests locally on my hardware and I noticed failures on `nrf52840dk_nrf52840` with following tests:
1. `nrf/tests/drivers/flash_patch/flash_patch.flash_patch_off`
2. `nrf/tests/drivers/fprotect/positive/drivers.fprotect.positive`
3. `nrf/tests/lib/hw_unique_key/lib.hw_unique_key`

What is more interesting - I tried to run those tests on current main (without your changes in this PR) and those tests failed too. So, it is probably not connected with migration to new Ztest API. Do I need some special board revision or hardware setup to pass those tests? 

I'm working with:
- OS: Linux Ubuntu
- Toolchain: zephyr-sdk-0.15.0
- nrf52840dk_nrf52840 board revision: 2.0.1

To reproduce
---
`nrf/tests/drivers/flash_patch/flash_patch.flash_patch_off`
Twister command:
```
./zephyr/scripts/twister -vv --inline-logs -p nrf52840dk_nrf52840 --device-testing --device-serial /dev/ttyACM0 -T nrf/tests/drivers/flash_patch -s flash_patch.flash_patch_off
```
Platform output:
```
*** Booting Zephyr OS build v3.2.99-ncs2-3160-g74c7bca86e6a ***
Running TESTSUITE test_flash_patch
===================================================================
START - test_flash_patch1
Assertion failed at WEST_TOPDIR/nrf/tests/drivers/flash_patch/src/main.c:18: test_flash_patch_test_flash_patch1: (NRF_UICR->DEBUGCTRL not equal to 0xFFFFFFFF)
Flash patch protection wrongly applied (0xffff00ff).
FAIL - test_flash_patch1 in 0.020 seconds
===================================================================
TESTSUITE test_flash_patch failed.
------ TESTSUITE SUMMARY START ------
SUITE FAIL -   0.00% [test_flash_patch]: pass = 0, fail = 1, skip = 0, total = 1 duration = 0.020 seconds
- FAIL - [test_flash_patch.test_flash_patch1] duration = 0.020 seconds
------ TESTSUITE SUMMARY END ------
===================================================================
RunID: 812e8dc2d7069c388f354f27ee13c73d
PROJECT EXECUTION FAILED
```
---
`nrf/tests/drivers/fprotect/positive/drivers.fprotect.positive`
Twister command:
```
./zephyr/scripts/twister -vv --inline-logs -p nrf52840dk_nrf52840 --device-testing --device-serial /dev/ttyACM0 -T nrf/tests/drivers/fprotect/positive
```
Platform output:
```
*** Booting Zephyr OS build v3.2.99-ncs2-3160-g74c7bca86e6a ***
Attempting to boot slot 0.

Attempting to boot from address 0x9000.
Verifying signature against key 0.
Hash: 0x39...c8

Firmware signature verified.
Firmware version 1

Setting monotonic counter (version: 1, slot: 0)

Booting (0x9000).

*** Booting Zephyr OS build v3.2.99-ncs2-3160-g74c7bca86e6a ***
Running TESTSUITE test_fprotect_positive
===================================================================
START - test_1_flash_write
PASS - test_1_flash_write in 0.001 seconds
===================================================================
START - test_2_flash_read
Assertion failed at WEST_TOPDIR/nrf/tests/drivers/fprotect/positive/src/main.c:45: test_fprotect_positive_test_2_flash_read: (read_data_after[i] not equal to write_data[i])
Expected:'e' got 'A' at positition 1
FAIL - test_2_flash_read in 0.020 seconds
===================================================================
START - test_3_flash_read_protected
PASS - test_3_flash_read_protected in 0.001 seconds
===================================================================
TESTSUITE test_fprotect_positive failed.
------ TESTSUITE SUMMARY START ------
SUITE FAIL -  66.67% [test_fprotect_positive]: pass = 2, fail = 1, skip = 0, total = 3 duration = 0.022 seconds
- PASS - [test_fprotect_positive.test_1_flash_write] duration = 0.001 seconds
- FAIL - [test_fprotect_positive.test_2_flash_read] duration = 0.020 seconds
- PASS - [test_fprotect_positive.test_3_flash_read_protected] duration = 0.001 seconds
------ TESTSUITE SUMMARY END ------
===================================================================
RunID: c039c6176bcf1e168958764d99708253
PROJECT EXECUTION FAILED
```

---

`nrf/tests/lib/hw_unique_key/lib.hw_unique_key`
Twister command:
```
./zephyr/scripts/twister -vv --inline-logs -p nrf52840dk_nrf52840 --device-testing --device-serial /dev/ttyACM0 -T nrf/tests/lib/hw_unique_key
```
Platform output:
```
*** Booting Zephyr OS build v3.2.99-ncs2-3160-g74c7bca86e6a ***
Running TESTSUITE test_hw_unique_key
===================================================================
START - test_hw_unique_key1
This test requires a full chip erase to be rerun.
*** Booting Zephyr OS build v3.2.99-ncs2-3160-g74c7bca86e6a ***
Running TESTSUITE test_hw_unique_key
===================================================================
START - test_hw_unique_key1
Assertion failed at WEST_TOPDIR/nrf/tests/lib/hw_unique_key/src/main.c:151: do_key_test: (Reached unreachable code)
An error occurred.
FAIL - test_hw_unique_key1 in 0.013 seconds
===================================================================
TESTSUITE test_hw_unique_key failed.
------ TESTSUITE SUMMARY START ------
SUITE FAIL -   0.00% [test_hw_unique_key]: pass = 0, fail = 1, skip = 0, total = 1 duration = 0.013 seconds
- FAIL - [test_hw_unique_key.test_hw_unique_key1] duration = 0.013 seconds
------ TESTSUITE SUMMARY END ------
===================================================================
RunID: 82243c151b001b437377a44cbfac4457
PROJECT EXECUTION FAILED
```

**Comment by oyvindronningstad** [comment]
@gopiotr Thanks for checking. All those tests pass on my desk though. Can you double check your setup?

I'm using Ubuntu, zephyr-sdk-0.16.0, and PCA10056 v3.0.0.

**Comment by gopiotr** [comment]
@oyvindronningstad thank you for letting me know. I try to test it on newer board revision (the same like your - v3.0.0) and it works better now. `nrf/tests/drivers/fprotect/positive/drivers.fprotect.positive` and `nrf/tests/lib/hw_unique_key/lib.hw_unique_key` pass properly, but `nrf/tests/drivers/flash_patch/flash_patch.flash_patch_off` still fails. 

I run it by command:
```
./zephyr/scripts/twister -vv --inline-logs -p nrf52840dk_nrf52840 --device-testing --device-serial /dev/ttyACM0 -T nrf/tests/drivers/flash_patch -s flash_patch.flash_patch_off
```
And I get consistently the same failure:
```
2023-04-21 14:33:33,347 - twister - DEBUG - Flash command: ['ninja', '-C', '/home/kagi/ncs/nrf/twister-out/nrf52840dk_nrf52840/flash_patch.flash_patch_off', 'flash']
2023-04-21 14:33:44,861 - twister - DEBUG - DEVICE: *** Booting Zephyr OS build v3.2.99-ncs2-3160-g74c7bca86e6a ***
2023-04-21 14:33:44,865 - twister - DEBUG - DEVICE: Running TESTSUITE test_flash_patch
2023-04-21 14:33:44,870 - twister - DEBUG - DEVICE: ===================================================================
2023-04-21 14:33:44,873 - twister - DEBUG - DEVICE: START - test_flash_patch1
2023-04-21 14:33:44,874 - twister - DEBUG - DEVICE:
2023-04-21 14:33:44,890 - twister - DEBUG - DEVICE: Assertion failed at WEST_TOPDIR/nrf/tests/drivers/flash_patch/src/main.c:18: test_flash_patch_test_flash_patch1: (NRF_UICR->DEBUGCTRL not equal to 0xFFFFFFFF)
2023-04-21 14:33:44,894 - twister - DEBUG - DEVICE: Flash patch protection wrongly applied (0xffff00ff).
2023-04-21 14:33:44,894 - twister - DEBUG - DEVICE:
2023-04-21 14:33:44,899 - twister - DEBUG - DEVICE: FAIL - test_flash_patch1 in 0.020 seconds
2023-04-21 14:33:44,904 - twister - DEBUG - DEVICE: ===================================================================
2023-04-21 14:33:44,908 - twister - DEBUG - DEVICE: TESTSUITE test_flash_patch failed.
```
Could you please show your output from this test? Does it require some special board preparing before test?

**Review by gopiotr — APPROVED** [review]
@oyvindronningstad - I believe that this have to be problem with my hardware or software setup and that's why I still have a problem with this `flash_patch.flash_patch_off` test. I think that this is not a blocker for this PR. Thank you for your effort!

**Pull request merged**